src/mjolnir/graphfilter.cc

```diff
@@ -219,7 +393,359 @@ void FilterTiles(GraphReader& reader,
+        return;
+      }
+
+      // walk in the correct direction.
+      uint64_t wayid = edgeinfo.wayid();
+      if (!Aggregate(id, reader, shape, en, from_node, wayid, isos, isForward)) {
+        LOG_WARN("ValidateData - failed to validate node.  Will not aggregate.");
+        std::cout << "End node: " << directededge->endnode() << " WayId: " << edgeinfo.wayid()
+                  << std::endl;
+
+        if (wayid == 0) { // This edge has special attributes, we can't aggregate
+          no_agg_ways.insert(edgeinfo.wayid());
+        }
+
+        processed_nodes.insert(directededge->endnode()); // turn off so that we don't fail
+      } else if (isos.size() > 1) {                      // in diff country
+        processed_nodes.insert(directededge->endnode());
+      }
+    }
+  }
+}
+
+void AggregateTiles(GraphReader& reader, std::unordered_map<GraphId, GraphId>& old_to_new) {
+
+  LOG_INFO("Validating edges for aggregation");
+  // Iterate through all tiles in the local level
+  auto local_tiles = reader.GetTileSet(TileHierarchy::levels().back().level);
+  for (const auto& tile_id : local_tiles) {
+    // Get the graph tile. Read from this tile to create the new tile.
+    graph_tile_ptr tile = reader.GetGraphTile(tile_id);
+    assert(tile);
+
+    std::unordered_set<GraphId> processed_nodes;
+    std::unordered_set<uint64_t> no_agg_ways;
+    processed_nodes.reserve(tile->header()->nodecount());
+    no_agg_ways.reserve(tile->header()->directededgecount());
+
+    GraphId nodeid = GraphId(tile_id.tileid(), tile_id.level(), 0);
+    for (uint32_t i = 0; i < tile->header()->nodecount(); ++i, ++nodeid) {
+      const NodeInfo* nodeinfo = tile->node(i);
+      uint32_t idx = nodeinfo->edge_index();
+      for (uint32_t j = 0; j < nodeinfo->edge_count(); j++, idx++) {
+        const DirectedEdge* directededge = tile->directededge(idx);
+        if (processed_nodes.find(nodeid) == processed_nodes.end()) {
+          GraphId en = directededge->endnode();
+          std::list<PointLL> shape;
+          // check if we can aggregate the edges at this node.
+          ValidateData(reader, shape, en, processed_nodes, no_agg_ways, nodeid, tile, directededge);
+        }
+      }
+    }
+
+    if (reader.OverCommitted()) {
+      reader.Trim();
+    }
+
+    // Now loop again double checking the ways.
+    nodeid = GraphId(tile_id.tileid(), tile_id.level(), 0);
+    for (uint32_t i = 0; i < tile->header()->nodecount(); ++i, ++nodeid) {
+      const NodeInfo* nodeinfo = tile->node(i);
+      uint32_t idx = nodeinfo->edge_index();
+      for (uint32_t j = 0; j < nodeinfo->edge_count(); j++, idx++) {
+        const DirectedEdge* directededge = tile->directededge(idx);
+        if (no_agg_ways.find(tile->edgeinfo(directededge).wayid()) != no_agg_ways.end()) {
+          processed_nodes.insert(directededge->endnode());
+        }
+      }
+    }
+
+    if (reader.OverCommitted()) {
+      reader.Trim();
+    }
```

Comment:
once is enough inside this outer loop no (s. line 533)? maybe we can move it to the end of the loop?

Reply:
oh you actually do it at the very end, so I guess both of these can go

Reply:
ty